File: _tests/validate.rb

=begin
checks our data for syntax errors and compliance with the schema
=end

require "kwalify"
require "yaml"

def validate(kind)
  validator = Kwalify::Validator.new(
                YAML.load_file(File.join("_tests", "schemas",
                                         "#{kind.chop}.yaml")))

  Dir[File.join(kind, "*.yaml")].each do |filename|
    errors = validator.validate(YAML.load_file(filename))
    if errors && !errors.empty?
      puts filename
      errors.each { |error| puts "\t[#{error.path}] #{error.message}" }
    end
  end
end

validate "bills"
validate "committee_reports"
validate "mps"
validate "plenary_sittings"

File: _schemas/tests/validate.py

"""Checks our data for syntax errors and compliance with the
schema.
"""

from pathlib import Path
import sys

from jsonschema import (
    Draft4Validator as Validator,
    FormatChecker)
import yaml


def main(folder):
    exit_status = 0

    with (Path('_schemas')/'{filename}.yaml'.format(
            filename=folder.rstrip('s/'))).open() as f:
        validator = Validator(yaml.load(f), format_checker=FormatChecker(
            ['date']))

    for item in Path(folder).iterdir():
        with item.open() as f:
            errors = validator.iter_errors(yaml.load(f))
            if errors:
                exit_status = 1
                for error in errors:
                    print('{item}: [{field}] {message}'.format(
                        item=item, field='/'.join(map(str, error.path)),
                        message=error.message))

    return exit_status


if __name__ == '__main__':
    sys.exit(main(sys.argv[1]))
